src/wui/http/http_client.c

```diff
@@ -7,121 +7,503 @@
     const char *header_plus_data = get_update_str(header);
 
-    uint16_t len = strlen(header_plus_data);
-    /* push to buffer */
-    err_t error = tcp_write(testpcb, header_plus_data, len, TCP_WRITE_FLAG_COPY);
+    req_len = strlen(header_plus_data);
 
-    if (error) {
-
-        return 1;
+    if ((req_len < 0) || (req_len > 0xFFFF)) {
+        return ERR_VAL;
     }
-
-    /* now send */
-    error = tcp_output(testpcb);
-    if (error) {
-
-        return 1;
+    /* alloc state and request in one block */
+    alloc_len = sizeof(httpc_state_t);
+    mem_alloc_len = (mem_size_t)alloc_len;
+    if ((mem_alloc_len < alloc_len) || (req_len + 1 > 0xFFFF)) {
+        return ERR_VAL;
     }
-    return 0;
-}
-
-/* connection established callback, err is unused and only return 0 */
-err_t connectCallback(void *arg, struct tcp_pcb *tpcb, err_t err) {
-
-    tcp_send_packet();
-    return 0;
-}
 
-void buddy_http_client_init() {
+    req = (httpc_state_t *)mem_malloc((mem_size_t)alloc_len);
+    if (req == NULL) {
+        return ERR_MEM;
+    }
+    memset(req, 0, sizeof(httpc_state_t));
+    req->timeout_ticks = HTTPC_POLL_TIMEOUT;
+    req->request = pbuf_alloc(PBUF_RAW, (u16_t)(req_len + 1), PBUF_RAM);
+    if (req->request == NULL) {
+        httpc_free_state(req);
+        return ERR_MEM;
+    }
+    if (req->request->next != NULL) {
+        /* need a pbuf in one piece */
+        httpc_free_state(req);
+        return ERR_MEM;
+    }
+    req->hdr_content_len = HTTPC_CONTENT_LEN_INVALID;
+    req->pcb = altcp_new();
+    if (req->pcb == NULL) {
+        httpc_free_state(req);
+        return ERR_MEM;
+    }
+    req->remote_port = CLIENT_PORT_NO;
+    altcp_arg(req->pcb, req);
+    altcp_recv(req->pcb, httpc_tcp_recv);
+    altcp_err(req->pcb, httpc_tcp_err);
+    altcp_poll(req->pcb, httpc_tcp_poll, HTTPC_POLL_INTERVAL);
+    altcp_sent(req->pcb, httpc_tcp_sent);
 
-    ip_addr_t ip;
-    ip.addr = eeprom_get_var(EEVAR_CONNECT_IP).ui32;
-    /* create the control block */
-    testpcb = tcp_new();
-    tcp_arg(testpcb, &data);
+    /* set up request buffer */
+    req_len2 = strlcpy((char *)req->request->payload, header_plus_data, req_len + 1);
+    if (req_len2 != req_len) {
+        httpc_free_state(req);
+        return ERR_VAL;
+    }
 
-    /* register callbacks with the pcb */
-    tcp_err(testpcb, tcpErrorHandler);
-    //tcp_recv(testpcb, tcpRecvCallback);
-    tcp_sent(testpcb, tcpSendCallback);
-    /* now connect */
-    tcp_connect(testpcb, &ip, 9000, connectCallback);
+    req->recv_fn = data_received_fun;
+    tcp_connect(req->pcb, &host_ip4, 9000, httpc_tcp_connected);
+    return ERR_OK;
 }
 
 void buddy_http_client_loop() {
 
     if (!init_tick) {
-        client_interval = HAL_GetTick();
+        client_interval = xTaskGetTickCount();
         init_tick = true;
     }
-
-    if (netif_ip4_addr(&eth0)->addr != 0 && ((HAL_GetTick() - client_interval) > CLIENT_CONNECT_DELAY)) {
+    xTaskGetTickCount();
```

Comment:
For what purpose is this xTaskG... call?

Reply:
thanks! it shouldn't be there